src/Microsoft.OData.Core/Json/PooledByteBufferWriter.cs

```diff
@@ -0,0 +1,221 @@
+        /// <param name="count"></param>
+        public void Advance(int count)
+        {
+            Debug.Assert(this.rentedBuffer != null);
+            Debug.Assert(count >= 0);
+            Debug.Assert(this.index <= this.rentedBuffer.Length - count);
+            this.index += count;
+        }
+
+        /// <summary>
+        /// <inheritdoc/>
+        /// </summary>
+        /// <param name="sizeHint"></param>
+        /// <returns></returns>
+        public Memory<byte> GetMemory(int sizeHint = MinimumBufferSize)
+        {
+            CheckAndResizeBuffer(sizeHint);
+            return this.rentedBuffer.AsMemory(index);
+        }
+
+        /// <summary>
+        /// <inheritdoc/>
+        /// </summary>
+        /// <param name="sizeHint"></param>
+        /// <returns></returns>
+        public Span<byte> GetSpan(int sizeHint = MinimumBufferSize)
+        {
+            if (sizeHint == 0)
+            {
+                sizeHint = MinimumBufferSize;
+            }
+
+            CheckAndResizeBuffer(sizeHint);
+            return this.rentedBuffer.AsSpan(index);
+        }
+
+        /// <summary>
+        /// Clears the data written to the underlying buffer.
+        /// </summary>
+        public void Clear()
+        {
+            ClearHelper();
+        }
+
+        /// <summary>
+        /// Returns the rented memory back to the pool.
+        /// </summary>
+        public void Dispose()
+        {
+            if (this.rentedBuffer == null)
+            {
+                return;
+            }
+
+            ClearHelper();
+            byte[] toReturn = this.rentedBuffer;
+            this.rentedBuffer = null;
+            ArrayPool<byte>.Shared.Return(toReturn);
+        }
+
+        private void ClearHelper()
+        {
+            Debug.Assert(this.rentedBuffer != null);
+            Debug.Assert(this.index <= this.rentedBuffer.Length);
+
+            this.rentedBuffer.AsSpan(0, index).Clear();
+            this.index = 0;
+        }
+
+        private void CheckAndResizeBuffer(int sizeHint)
+        {
+            Debug.Assert(rentedBuffer != null);
+            Debug.Assert(sizeHint > 0);
+
+            int currentLength = this.rentedBuffer.Length;
+            int availableSpace = currentLength - this.index;
+
+            // If we've reached ~1GB written, grow to the maximum buffer
+            // length to avoid incessant minimal growths causing perf issues.
+            if (this.index >= MaximumBufferSize / 2)
+            {
+                sizeHint = Math.Max(sizeHint, MaximumBufferSize - currentLength);
+            }
+
+            if (sizeHint > availableSpace)
+            {
+                int growBy = Math.Max(sizeHint, currentLength);
+
+                int newSize = currentLength + growBy;
+
+                if ((uint)newSize > MaximumBufferSize)
+                {
+                    newSize = currentLength + sizeHint;
+                    if ((uint)newSize > MaximumBufferSize)

```

Comment:
I don't think I understand what is happening here. Maybe you could add some comments like you did in the `if` block before this...

Reply:
I added comments.